yt/yt/client/table_client/composite_compare.cpp

```diff
@@ -276,4 +308,112 @@ TFingerprint CompositeFarmHash(TYsonStringBuf value)
 
 ////////////////////////////////////////////////////////////////////////////////
 
+std::optional<TYsonString> TruncateCompositeValue(TYsonStringBuf value, i64 size)
+{
+    YT_ASSERT(value.GetType() == EYsonType::Node);
+
+    if (size <= 0) {
+        return {};
+    }
+
+    TMemoryInput valueIn(value.AsStringBuf());
+    TYsonPullParser valueParser(&valueIn, EYsonType::Node);
+
+    TString truncatedYson;
+    TStringOutput output(truncatedYson);
+    output.Reserve(std::min(size, std::ssize(value.AsStringBuf())));
+    TCheckedInDebugYsonTokenWriter writer(&output);
+
+    i64 unclosedListCount = 0;
+    bool nonEmptyResult = false;
+
+    for (auto remainingBytes = size;;) {
+        const auto item = valueParser.Next();
+
+        // Both braces of the outmost list are not considered to be inside an enclosing list.
+        bool isInsideList = unclosedListCount - (item.GetType() == EYsonItemType::EndList);
+        auto resultingItemSize = GetMinResultingSize(item, /*isInsideList*/ isInsideList);
+
+        if (resultingItemSize > remainingBytes) {
+            break;
+        }
+
+        bool canBeTruncatedFurther = true;
+
+        switch (item.GetType()) {
+            case EYsonItemType::BeginList:
+                ++unclosedListCount;
+                writer.WriteBeginList();
+                break;
+            case EYsonItemType::EndList:
+                --unclosedListCount;
+                writer.WriteEndList();
+                break;
+            case EYsonItemType::EntityValue:
+                writer.WriteEntity();
+                break;
+            case EYsonItemType::Int64Value:
+                writer.WriteBinaryInt64(item.UncheckedAsInt64());
+                break;
+            case EYsonItemType::Uint64Value:
+                writer.WriteBinaryUint64(item.UncheckedAsUint64());
+                break;
+            case EYsonItemType::DoubleValue:
+                writer.WriteBinaryDouble(item.UncheckedAsDouble());
+                break;
+            case EYsonItemType::BooleanValue:
+                writer.WriteBinaryBoolean(item.UncheckedAsBoolean());
+                break;
+            case EYsonItemType::StringValue: {
+                auto truncatedString = item.UncheckedAsString().Trunc(remainingBytes - resultingItemSize);
+                writer.WriteBinaryString(truncatedString);
+                // This is a slight overestimation, since a smaller string might have a shorter varint part storing its length.
+                resultingItemSize += truncatedString.size();
+                break;
+            }
+            // Maps and attributes are not comparable.
+            // However, we can store everything up to this point into the truncated string.
+            case EYsonItemType::BeginAttributes:
+            case EYsonItemType::EndAttributes:
+            case EYsonItemType::BeginMap:
+            case EYsonItemType::EndMap:
+            case EYsonItemType::EndOfStream:
```

Comment:
Are BeginAttribtes, EndAttributes and EndOfStream even possible in this context?

Reply:
EndOfStream is definitely possible from the pull parser. Attributes probably won't appear in table data, but this is a generic function, so why not?